wpilibj/src/main/java/edu/wpi/first/wpilibj/experimental/controller/PIDController.java

```diff
@@ -0,0 +1,581 @@
+   * @param maximumInput the maximum value expected from the output
+   */
+  public void setInputRange(double minimumInput, double maximumInput) {
+    m_thisMutex.lock();
+    try {
+      m_minimumInput = minimumInput;
+      m_maximumInput = maximumInput;
+      m_inputRange = maximumInput - minimumInput;
+    } finally {
+      m_thisMutex.unlock();
+    }
+
+    setReference(getReference());
+  }
+
+  /**
+   * Sets the minimum and maximum values to write.
+   *
+   * @param minimumOutput the minimum value to write to the output
+   * @param maximumOutput the maximum value to write to the output
+   */
+  public void setOutputRange(double minimumOutput, double maximumOutput) {
+    m_thisMutex.lock();
+    try {
+      m_minimumOutput = minimumOutput;
+      m_maximumOutput = maximumOutput;
+    } finally {
+      m_thisMutex.unlock();
+    }
+  }
+
+  /**
+   * Set the absolute error which is considered tolerable for use with atReference().
+   *
+   * @param tolerance Absolute error which is tolerable.
+   */
+  public void setAbsoluteTolerance(double tolerance) {
+    setAbsoluteTolerance(tolerance, Double.POSITIVE_INFINITY);
+  }
+
+  /**
+   * Set the absolute error which is considered tolerable for use with atReference().
+   *
+   * @param tolerance      Absolute error which is tolerable.
+   * @param deltaTolerance Change in absolute error per second which is tolerable.
+   */
+  public void setAbsoluteTolerance(double tolerance, double deltaTolerance) {
+    m_thisMutex.lock();
+    try {
+      m_toleranceType = Tolerance.kAbsolute;
+      m_tolerance = tolerance;
+      m_deltaTolerance = deltaTolerance;
+    } finally {
+      m_thisMutex.unlock();
+    }
+  }
+
+  /**
+   * Set the percent error which is considered tolerable for use with atReference().
+   *
+   * @param tolerance Percent error which is tolerable.
+   */
+  public void setPercentTolerance(double tolerance) {
+    setPercentTolerance(tolerance, Double.POSITIVE_INFINITY);
+  }
+
+  /**
+   * Set the percent error which is considered tolerable for use with atReference().
+   *
+   * @param tolerance      Percent error which is tolerable.
+   * @param deltaTolerance Change in percent error per second which is tolerable.
+   */
+  public void setPercentTolerance(double tolerance, double deltaTolerance) {
+    m_thisMutex.lock();
+    try {
+      m_toleranceType = Tolerance.kPercent;
+      m_tolerance = tolerance;
+      m_deltaTolerance = deltaTolerance;
+    } finally {
+      m_thisMutex.unlock();
+    }
+  }
+
+  /**
+   * Returns the difference between the reference and the measurement.
+   *
+   * @return The error.
+   */
+  public double getError() {
+    m_thisMutex.lock();
+    try {
+      return getContinuousError(m_currError);
+    } finally {
+      m_thisMutex.unlock();
+    }
+  }
+
+  /**
+   * Returns the change in error per second.
+   */
+  public double getDeltaError() {
+    double error = getError();
```

Comment:
Why outside of lock?

Reply:
Because getError() locks the mutex as well, and we want to avoid recursive locking for performance reasons. It might not matter in Java because all locks are recursive anyway, but it does matter in C++, so we do the same here.